### DISCUSSION TITLE
JdbcDatabaseContainer overrides user and password with values from testcontainer

### DISCUSSION THREAD

**Discussion by gusega** [testcontainers/testcontainers-java#7724] [Q&A]
HI all,

I was trying to make flyway migration work with testcontainers but they wouldn't. 
It appears that [JdbcDatabaseContainer](https://github.com/testcontainers/testcontainers-java/blob/main/modules/jdbc/src/main/java/org/testcontainers/containers/JdbcDatabaseContainer.java#L233) overrides explicitly provided user and password. 

The [default oracle container](https://github.com/testcontainers/testcontainers-java/blob/main/modules/oracle-xe/src/main/java/org/testcontainers/containers/OracleContainer.java#L62) sets username and password to `test`/`test`. 

In my configuration I explicitly provided `spring.flyway.user=system; spring.flyway.password=test`. But the result connection will use the container's default `test`/`test` that has no privileges. 

In my opinion, if user and password is explicitly provided in `info` to `JdbcDatabaseContainer.createConnection()` they should be used, not the defaults from the container. 

I probably should create my own container implementation instead of using default one.

**Comment by eddumelendez** [comment]
You can change the default values provided by OracleContainer using withUsername, withPassword. When using spring boot < 3.1 you can use `@DynamicPropertySource` and set the jdbc url, username and password properties with the values provided by Testcontainers or if using spring boot >= 3.1 then just add the `@ServiceConnection` to the OracleContainer.

**Comment by gusega** [comment]
`gvenzl/oracle-xe` image tries to create a new user if I set `withUsername()` or `withPassword()`.
 
What do you mean by:
> set the jdbc url, username and password properties with the values provided by Testcontainers

?

**Reply by eddumelendez** [reply]
Check [this guide](https://testcontainers.com/guides/testing-spring-boot-rest-api-using-testcontainers/)

```
@DynamicPropertySource
static void configureProperties(DynamicPropertyRegistry registry) {
  registry.add("spring.datasource.url", postgres::getJdbcUrl);
  registry.add("spring.datasource.username", postgres::getUsername);
  registry.add("spring.datasource.password", postgres::getPassword);
}
```

**Comment by gusega** [comment]
ah, I got what you mean. Instead of using jdbc connection, create the container in the test and then provide jdbc url and username/password to spring properties. That's actually quite nice. 

But the issue is in the `OracleContainer` and in the `createConnection()` method of `JdbcDatabaseContainer` as I described above.

For now I neded up with copy-pasting `OracleContainer` with removing the 
```
        withEnv("APP_USER", username);
        withEnv("APP_USER_PASSWORD", password);
```
parts from `configure()` 
and overriding the `createConnection(String queryString, Properties info)`
  without the 
  ```
        properties.put("user", this.getUsername());
        properties.put("password", this.getPassword());

  ``` 
  
  part.
  
 And now it works.

**Reply by eddumelendez** [reply]
Please, use the thread to reply a given message. I am not sure I follow, but you should not use the `createConnection`. Using the pre-configure database modules should be as simple as

```java
OracleContainer oracle = new OracleContainer("<insert-your-image-and-tag>");
var jdbcUrl = oracle.getJdbcUrl();
var username = oracle.getUsername(); // test
var password = oracle.getPasword();  // test
```

if other credentials are needed then

```java
OracleContainer oracle = new OracleContainer("<insert-your-image-and-tag>")
    .withUsername("testuser")
    .withPassword("testpassword");
var jdbcUrl = oracle.getJdbcUrl();
var username = oracle.getUsername(); // testuser
var password = oracle.getPasword();  // testpassword
```

**Reply by gusega** [reply]
@eddumelendez  right thanks.
let me tell the story again in a more structured way. Our set up is the following:

1. an empty oracle from testcontainers with only `system` user
2. run flyway migrations as `system` user. Migrations create schema and other users including `app_user`
3. run the application (tests) with the `app_user` 


I ended up with [this OracleContainer](https://gist.github.com/gusega/4515fe143a60774904fc6c8d4c3d4362). I copy pasted  flyway's `oracle-xe` and adjusted the pieces mentioned above.

It allows to use different user for different connections: migration as `system`, run application as `app_user`. The container on start has only `system` user, which is equivalent to running container with [`docker run -d -p 1521:1521 -e ORACLE_PASSWORD=<your password> gvenzl/oracle-xe`](https://hub.docker.com/r/gvenzl/oracle-xe#:~:text=docker%20run%20%2Dd%20%2Dp%201521%3A1521%20%2De%20ORACLE_PASSWORD%3D%3Cyour%20password%3E%20gvenzl/oracle%2Dxe).
